File: app/src/main/java/com/bzh/dytt/data/network/NetworkResource.java
package com.bzh.dytt.data.network;


import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.Observer;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import com.bzh.dytt.AppExecutors;

public abstract class NetworkResource<ResultType, RequestType> {

    private final AppExecutors mAppExecutors;

    private MediatorLiveData<Resource<ResultType>> mResult = new MediatorLiveData<>();

    @MainThread
    public NetworkResource(AppExecutors appExecutors) {
        mAppExecutors = appExecutors;

        mResult.setValue(Resource.<ResultType>loading(null));

        Runnable diskRunnable = new Runnable() {
            @Override
            public void run() {
                onPreProcess();
                Runnable mainRunnable = new Runnable() {
                    @Override
                    public void run() {
                        final LiveData<ApiResponse<RequestType>> apiResponse = createCall();
                        mResult.addSource(apiResponse, new Observer<ApiResponse<RequestType>>() {
                            @Override
                            public void onChanged(@Nullable final ApiResponse<RequestType> response) {
                                mResult.removeSource(apiResponse);

                                if (response.isSuccessful()) {
                                    mAppExecutors.diskIO().execute(new Runnable() {
                                        @Override
                                        public void run() {
                                            final ResultType newData = saveCallResult(processResponse(response));
                                            mAppExecutors.mainThread().execute(new Runnable() {
                                                @Override
                                                public void run() {
                                                    mResult.setValue(Resource.success(newData));
                                                }
                                            });
                                        }
                                    });
                                } else {
                                    onFetchFailed();
                                    mResult.setValue(Resource.<ResultType>error(response.errorMessage, null));
                                }
                            }
                        });
                    }
                };
                mAppExecutors.mainThread().execute(mainRunnable);
            }
        };
        mAppExecutors.diskIO().execute(diskRunnable);
    }

    @WorkerThread
    protected void onPreProcess() {
    }

    @WorkerThread
    protected RequestType processResponse(ApiResponse<RequestType> response) {
        return response.body;
    }

    // Called to insertItem the mResult of the API response into the database
    @WorkerThread
    protected abstract ResultType saveCallResult(@NonNull RequestType item);

    // Called to create the API call.
    @NonNull
    @MainThread
    protected abstract LiveData<ApiResponse<RequestType>> createCall();

    // Called when the fetch fails. The child class may want to reset components
    // like rate limiter.
    @MainThread
    protected void onFetchFailed() {
    }

    // returns a LiveData that represents the resource, implemented
    // in the base class.
    public final LiveData<Resource<ResultType>> getAsLiveData() {
        return mResult;
    }
}

File: app/src/main/java/com/bzh/dytt/data/network/NetworkResource.kt
package com.bzh.dytt.data.network

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MediatorLiveData
import android.support.annotation.MainThread
import android.support.annotation.WorkerThread
import com.bzh.dytt.AppExecutors
import com.bzh.dytt.data.Resource

@MainThread
abstract class NetworkResource<ResultType, RequestType>
constructor(private val appExecutors: AppExecutors) {

    private val result = MediatorLiveData<Resource<ResultType>>()

    init {
        result.value = Resource.loading(null)

        val diskRunnable = Runnable {
            onPreProcess()
            val mainRunnable = Runnable {
                val apiResponse = createCall()
                result.addSource(apiResponse) { response ->
                    result.removeSource(apiResponse)

                    when (response) {
                        is ApiSuccessResponse -> {
                            appExecutors.diskIO().execute {
                                val newData = saveCallResult(processResponse(response))
                                appExecutors.mainThread().execute {
                                    result.value = Resource.success(newData)
                                }
                            }
                        }
                        is ApiEmptyResponse -> {
                            result.value = Resource.success<ResultType>(null)
                        }
                        is ApiErrorResponse -> {
                            onFetchFailed()
                            result.value = Resource.error<ResultType>(response.errorMessage, null)
                        }
                    }
                }
            }
            appExecutors.mainThread().execute(mainRunnable)
        }
        appExecutors.diskIO().execute(diskRunnable)
    }

    @WorkerThread
    protected open fun onPreProcess() {
    }

    @WorkerThread
    protected open fun processResponse(response: ApiSuccessResponse<RequestType>) = response.body

    @WorkerThread
    protected abstract fun saveCallResult(item: RequestType): ResultType

    @MainThread
    protected abstract fun createCall(): LiveData<ApiResponse<RequestType>>

    @MainThread
    protected open fun onFetchFailed() {
    }

    fun getAsLiveData() = result
}